implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+    ///     the current party
+    pub(super) async fn new(vault: Arc<dyn XXVault>) -> Result<Handshake> {
+        // 1. generate a static key pair for this handshake and set it to s
+        let static_key = Self::generate_static_key(vault.clone()).await?;
+
+        // 2. generate an ephemeral key pair for this handshake and set it to e
+        let ephemeral_key = Self::generate_ephemeral_key(vault.clone()).await?;
+
+        // 3. initialize the handshake
+        // We currently don't use any payload for message 1
+        Self::new_with_keys(vault, static_key, ephemeral_key, vec![]).await
+    }
+
+    /// Initialize the handshake
+    async fn new_with_keys(
+        vault: Arc<dyn XXVault>,
+        static_key: KeyId,
+        ephemeral_key: KeyId,
+        message1_payload: Vec<u8>,
+    ) -> Result<Handshake> {
+        Ok(Handshake {
+            vault,
+            state: HandshakeState::new(static_key, ephemeral_key, message1_payload),
+        })
+    }
+
+    /// Import the k secret
+    async fn import_k_secret(&self, content: Vec<u8>) -> Result<KeyId> {
+        self.vault
+            .import_ephemeral_secret(Secret::new(content), Self::k_attributes())
+            .await
+    }
+
+    /// Import the ck secret
+    async fn import_ck_secret(&self, content: Vec<u8>) -> Result<KeyId> {
+        self.vault
+            .import_ephemeral_secret(Secret::new(content), Self::ck_attributes())
+            .await
+    }
+
+    /// Return the public key corresponding to a given key id
+    async fn get_public_key(&self, key_id: &KeyId) -> Result<PublicKey> {
+        self.vault.get_public_key(key_id).await
+    }
+
+    /// Compute a Diffie-Hellman key between a given key id and the other party public key
+    async fn dh(&self, key_id: &KeyId, public_key: &PublicKey) -> Result<KeyId> {
+        self.vault.ec_diffie_hellman(key_id, public_key).await
+    }
+
+    /// Compute two derived ck, and k keys based on existing ck and k keys + an optional
+    /// Diffie-Hellman key
+    async fn hkdf(&self, state: &mut HandshakeState, dh: Option<&KeyId>) -> Result<(KeyId, KeyId)> {
+        // if dh is not defined that means that we are doing the last step in the noise protocol and splitting
+        // an encryption key and a decryption key. In that case this means that we produce AES keys for both
+        // keys
+        let ck_attributes = if dh.is_none() {
+            Self::k_attributes()
+        } else {
+            Self::ck_attributes()
+        };
+
+        let hkdf_output = self
+            .vault
+            .hkdf_sha256(
+                &state.ck,
+                b"",
+                dh,
+                vec![ck_attributes, Self::k_attributes()],
+            )
+            .await?;
+
+        match hkdf_output.as_slice() {
+            [new_ck, new_k] => {
+                self.vault.delete_ephemeral_secret(state.ck.clone()).await?;
```

Comment:
Since those 2 fields are not `Option` (because of the previous discussion) it is is not possible. Also we don't yet consume `state` so things have to stay like that. One thing I simplified though was to remove return values for `hkdf` since we have a `mut state` that we can mutate right away to assign the new values for `ck` and `k`.